### PR TITLE
Added win_http_proxy and win_inet_proxy

### LINKED ISSUE #22663: Module proposal: win_wininet (configure IE proxy settings)
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME
win_wininet (proposed name)

##### ANSIBLE VERSION

```
2.2.1.0
```

##### SUMMARY
It hard/non-obvious to configure the ansible_user's "Internet Options" proxy configuration on Windows. This proposal suggests adding a new module to do that.

The potential benefit is that various other downloaders could then be made to work *implicitly* using wininet.dll. For example, .Net's System.Net.WebClient, or the "win_chocolatey" module.

##### OTHER NOTES

There are a number of **wrong** proposals online for configuring the IE proxy settings:
* Just set some specific registry keys (only works if the user account has had its IE profile initialized previously by running IE *at least once* -- which isn't helpful at all for automation)
* Set some registry keys and then "poke" wininet.dll to force a reload (same issue as last point)
* Use "netsh winhttp set proxy ..." (fundamentally misunderstands the difference between wininet and winhttp, which are different libraries used by different things in Windows)

A better solution is to initialize IE via wininet.dll itself. This can be done in PowerShell through inline C# with data marshaling for options content. Such a solution should be fairly easy for Ansible to deliver/manage.

I have a proof-of-concept for this as a binary written in C, which I have been using so far by first uploading the binary, executing it, and then cleaning it up, in multiple, separate Ansible tasks. (I can post the relevant source as soon as I get it cleared by my employer.)

### PR DESCRIPTION
**Pull request by jborean93**
##### SUMMARY
Adds 2 modules that can be used to configure proxy settings on Windows;

* win_http_proxy - Configures proxy settings for WinHTTP
* win_inet_proxy - Configures proxy settings for WinINet (IE)

While the input options are similar, WinINet offers a few extra options and WinHTTP can also import from WinINet. This made it difficult to have just 1 `win_proxy` module and so I created a module for each for a cleaner UI. Hopefully the documentation makes this clear as to that they do and link to each other.

Fixes https://github.com/ansible/ansible/issues/22663

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
win_http_proxy
win_inet_proxy

### CODE REVIEW

File: lib/ansible/modules/windows/win_http_proxy.ps1 [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+Add-CSharpType -References $win_http_invoke -AnsibleModule $module
+
+$actual_proxy = New-Object -TypeName Ansible.WinHttpProxy.WinHttpProxy
+
+$module.Diff.before = @{
+    proxy = $actual_proxy.Proxy
+    bypass = $actual_proxy.ProxyBypass
+}
+
+if ($source -eq "ie") {
```
Comment (johnboy2) [inline lib/ansible/modules/windows/win_http_proxy.ps1:221]: IIRC, the underlying "import from IE" logic provided by Microsoft silently ignores the autoconfig URL -- which seems to be a mistake to me. Do we want to propagate that behavior? I suspect a warning would be better.

File: lib/ansible/modules/windows/win_http_proxy.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: win_http_proxy
+version_added: '2.8'
+short_description: Manages proxy settings for WinHTTP
+description:
+- Used to set, remote, or import proxy settings for Windows HTTP Services
```
Comment (johnboy2) [inline lib/ansible/modules/windows/win_http_proxy.py]: `remote` -> `remove`?

File: lib/ansible/modules/windows/win_http_proxy.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  source:
+    description:
+    - Instead of manually specifying the I(proxy) and/or I(bypass), set this to
+      import the proxy from a set source like Internet Explorer.
+    - Using C(ie) will import the Internet Explorer proxy settings for the
+      current active network connection of the current user.
+    - This is like running C(netsh winhttp import proxy source=ie).
+    - The value is imported when the module runs and will not automatically
+      be updated if IE changes the values in the future. The module will have
+      to be run again to sync the latest changes.
```
Comment (johnboy2) [inline lib/ansible/modules/windows/win_http_proxy.py]: Nitpick - the term `The value` is used twice here with different meanings each time.

I suggest changing `...if IE changes the values...` to something like `...if the IE configuration changes...` to resolve the ambiguity.

File: lib/ansible/modules/windows/win_http_proxy.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    choices:
+    - ie
+    type: str
+notes:
+- This is not the same as the proxy settings set in Internet Explorer, also
+  known as C(WinINet), use the M(win_inet_proxy) module to manage that instead.
```
Comment (johnboy2) [inline lib/ansible/modules/windows/win_http_proxy.py]: Grammar nitpick... Quick fix here would be to swap the comma for a semi-colon.

The more detailed version is that this forms an invalid dependent clause the way it is written right now. A semi-colon would fix that, as would splitting off the `use the ...` content into its own sentence.

File: lib/ansible/modules/windows/win_http_proxy.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      connecting through any loopback address like C(127.0.0.1), C(localhost),
+      or the local hostname.
+    - Omit, set to null or an empty string/list to remove the bypass list.
+    - If this is set then I(proxy) must also be set.
+    type: list
+  proxy:
+    description:
+    - A string or dict that specifies the proxy to be set.
+    - If setting a string, should be in the form C(hostname), C(hostname:port),
+      or C(protocol=hostname:port).
```
Comment (johnboy2) [inline lib/ansible/modules/windows/win_http_proxy.py:35]: Should we document that default per-protocol ports get used if one isn't set explicitly? (This is one of those things that should be obvious, except to everyone for whom it isn't.)

File: lib/ansible/modules/windows/win_http_proxy.ps1 (hunk 1 above)
Comment (johnboy2) [reply lib/ansible/modules/windows/win_http_proxy.ps1:221]: (Admittedly, I haven't actually looked at that in a very long time, so I could be mis-remembering.)

File: lib/ansible/modules/windows/win_http_proxy.ps1 [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+        }
+    }
+    $proxy = $proxy_list -join ";"
+} elseif ($null -ne $proxy) {
+    $proxy = $proxy.ToString()
+}
+
+if ($bypass) {
+    $bypass = $bypass -join ';'
+}
```
Comment (johnboy2) [inline lib/ansible/modules/windows/win_http_proxy.ps1:64]: Should probably fail if `$bypass` is set to something but `$proxy` is not. (The `required_by` check isn't that strict.)

File: lib/ansible/modules/windows/win_inet_proxy.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: win_inet_proxy
+version_added: '2.8'
+short_description: Manages proxy settings for WinINet and Internet Explorer
+description:
+- Used to set, remote, or import proxy settings for Windows INet which includes
+  Internet Explorer.
```
Comment (johnboy2) [inline lib/ansible/modules/windows/win_inet_proxy.py]: `Used to set or remove proxy settings...`
* `remote` -> `remove`
* There is no importing IE settings here

File: lib/ansible/modules/windows/win_inet_proxy.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+- name: Remove any explicit proxies that are set
+  win_inet_proxy:
+    proxy: ''
+    bypass: ''
+
+# This should be done after setting the IE proxy with win_inet_proxy
+- name: Import IE proxy configuration to WinHTTP
+  win_http_proxy:
+    source: ie
+'''
```
Comment (johnboy2) [inline lib/ansible/modules/windows/win_inet_proxy.py:169]: Suggest adding a combined example that also uses win_credential.

**Review by johnboy2 — COMMENTED** [review]
Looks great overall!

I should be able to do some testing in my environment over the next week or so, so this is purely a code review for now.

File: lib/ansible/modules/windows/win_http_proxy.ps1 (hunk 1 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_http_proxy.ps1:221]: I couldn't find any way to actually set an auto config url in netsh/winhttp so I think it is a limitation of the latter? Could be wrong but the [WINHTTP_PROXY_INFO](https://docs.microsoft.com/en-us/windows/desktop/api/Winhttp/ns-winhttp-__unnamed_struct_3) structure does not have a field for this.

This sounds like more of a documentation thing with the potential of adding this in as a separate param if it is possible.

File: lib/ansible/modules/windows/win_http_proxy.py (hunk 2 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_http_proxy.py]: Thanks, fixed.

File: lib/ansible/modules/windows/win_http_proxy.py (hunk 3 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_http_proxy.py]: Thanks sounds a lot better.

File: lib/ansible/modules/windows/win_http_proxy.py (hunk 4 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_http_proxy.py]: I knew there was a reason I preferred Maths over English in school :)

File: lib/ansible/modules/windows/win_http_proxy.py (hunk 5 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_http_proxy.py:35]: Doesn't hurt, bound to get asked or someone would be interested to know.

File: lib/ansible/modules/windows/win_http_proxy.ps1 (hunk 6 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_http_proxy.ps1:64]: Not sure what you mean, `required_by` does exactly that and it's part of the integration tests.

File: lib/ansible/modules/windows/win_http_proxy.ps1 (hunk 1 above)
Comment (johnboy2) [reply lib/ansible/modules/windows/win_http_proxy.ps1:221]: Sorry, I guess I wasn't clear. The "mistake" I was referring to was the fact that netsh doing this type of import **_silently_** ignores the auto-config URL (IIRC).

I'm not surprised that there's no winhttp option for this, since that would imply winhttp.dll having a JS interpreter (which doesn't match well with its use-case). From that perspective, _ignoring_ the auto-config URL when doing an import is entirely reasonable on MS' part -- **_except_** if it is done silently. IMHO that should really be a warning.

File: lib/ansible/modules/windows/win_inet_proxy.py (hunk 8 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_inet_proxy.py:169]: Good idea, I forgot to actually test it out so I'm doing that right now.

File: lib/ansible/modules/windows/win_http_proxy.ps1 (hunk 1 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_http_proxy.ps1:221]: Hmm I might be missing something, it's early morning here. Just to double check you are saying that the netsh import doesn't use the auto config value at all? That's the assumption that I was under. The trouble with adding warnings to modules is that it will automatically display when being used. This case is more suited towards documentation where people reading the options should be aware that `source: ie` will just import the proxy URL and bypass list.

File: lib/ansible/modules/windows/win_http_proxy.ps1 (hunk 6 above)
Comment (johnboy2) [reply lib/ansible/modules/windows/win_http_proxy.ps1:64]: I can double-check, but I thought setting proxy to an empty string would break the check.

File: lib/ansible/modules/windows/win_http_proxy.ps1 (hunk 1 above)
Comment (johnboy2) [reply lib/ansible/modules/windows/win_http_proxy.ps1:221]: RE not used at all -- correct. (I'm still going from memory on that one, but that is how I remember it.) More specifically, if you import from IE, which _only_ is set to use a [valid] auto-config script, you'll get nothing whatsoever in the winhttp settings.

As far warning or doc, my _preference_ would be to do both, but I won't get up on a high horse about it either way. :-)

Comment (johnboy2) [reply lib/ansible/modules/windows/win_http_proxy.ps1:221]: (Should probably actually double-check my recollection on this first though. We'd all feel really silly if I'm remembering wrong.)

File: lib/ansible/modules/windows/win_http_proxy.ps1 (hunk 6 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_http_proxy.ps1:64]: Ah yes that's a good point, forgot that `proxy: ''` means no proxy at all. Will fix up.

File: lib/ansible/modules/windows/win_http_proxy.ps1 (hunk 1 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_http_proxy.ps1:221]: Cool so yes the import is only for explicit proxy URLs and bypass list. I'll update the docs to further clarify this but I don't think a warning is necessary. This is up to the user making sure they understand the behaviour and read the docs.

Comment (johnboy2) [reply lib/ansible/modules/windows/win_http_proxy.ps1:221]: Good enough for me.

**Comment by jborean93** [comment]
Thanks a lot for looking through this @johnboy2. I've made the changes requested and have also come across a bug in win_credential which I will fix outside of this PR. Hopefully I've covered everything you've brought up.

**Comment by jborean93** [comment]
Fix for win_credential https://github.com/ansible/ansible/pull/54695.

File: lib/ansible/modules/windows/win_http_proxy.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+options:
+  bypass:
+    description:
+    - A list of hosts that will bypass the set proxy when being accessed.
+    - Use C(<local>) to match hostnames that are not fully qualified domain
+      names. This is useful when needing to connect to intranet sites using
+      just the hostname.
+    - Use C(<-loopback>) to stop automatically bypassing the proxy when
+      connecting through any loopback address like C(127.0.0.1), C(localhost),
+      or the local hostname.
```
Comment (johnboy2) [inline lib/ansible/modules/windows/win_http_proxy.py]: Did some testing, but was _unable to validate_ this `<-loopback>` part, at least for winhttp.

In particular, on W2016, localhost and 127.0.0.1 destinations appear to ignore the proxy unconditionally, whether `<-loopback>` is in the bypass list or not. I'm also not sure my use of W2016 is a factor, since the MS docs seem to indicate it hasn't materially changed since Vista/2008. I have also not found any support for the idea that `<-loopback>` does anything in the MS docs I reviewed.

Note that I did [find a reference](https://docs.microsoft.com/en-us/windows/desktop/api/wininet/nf-wininet-internetopena) for this construct for wininet, but not for winhttp.

Is this a real thing?

**Comment by johnboy2** [comment]
Did some functional testing of these modules. Tried breaking them in a few different ways, but was mostly unsuccessful -- my tests all worked as expected _except_ for one. [See comment here](https://github.com/ansible/ansible/pull/54631/commits/bbf08126d2cf86ee7f262bce78bb264a9d91a89f#r271410503) for the one issue I found.

File: lib/ansible/modules/windows/win_http_proxy.ps1 (hunk 1 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_http_proxy.ps1:221]: I also found [WinHttpGetProxyForUrl](https://docs.microsoft.com/en-us/windows/desktop/api/winhttp/nf-winhttp-winhttpgetproxyforurl) which seems to be the mechanism for WinHTTP applications to support auto config urls. This indicates it's up to the application to look this up and set on their requests and not a system wide setting.

https://docs.microsoft.com/en-us/windows/desktop/WinHttp/winhttp-autoproxy-api for some more info.

File: lib/ansible/modules/windows/win_http_proxy.py (hunk 9 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_http_proxy.py]: It's definitely there for WinINet and I just assumed WinHTTP was the same but it seems like this isn't the case.

As you said I found the reference to `<-loopback>` in the `InternetOpen` function as well as https://blogs.msdn.microsoft.com/ieinternals/2013/10/11/understanding-web-proxy-configuration/. This then has a link to has a link to http://blogs.msdn.com/b/fiddler/archive/2011/02/10/fiddler-is-better-with-internet-explorer-9.aspx but like most MS articles I come across these days it is no longer valid.

Considering this was a feature added in IE9, your testing indicates it doesn't work for WinHTTP, and the lack of docs for this in WinHTTP I will just assume that it isn't valid and remove the references to it in the docs for `win_http_proxy`.

I'll also update the documentation for `win_inet_proxy` to say you need IE9 for this to work.

File: lib/ansible/modules/windows/win_http_proxy.ps1 (hunk 1 above)
Comment (johnboy2) [reply lib/ansible/modules/windows/win_http_proxy.ps1:221]: Yeah, your second link is pretty succinct on the matter:
> AutoProxy support is not fully integrated into the HTTP stack in WinHTTP. Before sending a request, the application must call ...

In that respect, your doc content changes look good.

**Comment by jborean93** [comment]
Thanks once again @johnboy2 for reviewing and testing this out. I've fixed the typos in the docs and removed mentions of `<-loopback>` for win_http_proxy.

File: lib/ansible/modules/windows/win_http_proxy.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    - If this is set then I(proxy) must also be set.
+    type: list
+  proxy:
+    description:
+    - A string or dict that specifies the proxy to be set.
+    - If setting a string, should be in the form C(hostname), C(hostname:port),
+      or C(protocol=hostname:port).
+    - If the port is undefined, the default port for the protocol in use is
+      used.
+    - If setting a dict, the keys relate are the protocol and the value is the
```
Comment (jhawkesworth) [inline lib/ansible/modules/windows/win_http_proxy.py]: Suggest rewording this sentence as 
"If setting a dict, the keys should be the protocol and the values should be the hostname and/or port for the protocol given in the key."

File: lib/ansible/modules/windows/win_inet_proxy.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+    - When ommited, the default LAN connection is used.
+    type: str
+  proxy:
+    description:
+    - A string or dict that specifies the proxy to be set.
+    - If setting a string, should be in the form C(hostname), C(hostname:port),
+      or C(protocol=hostname:port).
+    - If the port is undefined, the default port for the protocol in use is
+      used.
+    - If setting a dict, the keys relate are the protocol and the value is the
```
Comment (jhawkesworth) [inline lib/ansible/modules/windows/win_inet_proxy.py]: Again, suggest rewording this as suggested for win_http_proxy.

File: lib/ansible/modules/windows/win_inet_proxy.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+      hostname and/or port for the protocol specified.
+    - Valid protocols are C(http), C(https), C(ftp), and C(socks).
+    - Omit, set to null or an empty string to remove the proxy settings.
+notes:
+- This is not the same as the proxy settings set in WinHTTP through the
+  C(netsh) command. Use the M(win_http_proxy) module ot manage that instead.
```
Comment (jhawkesworth) [inline lib/ansible/modules/windows/win_inet_proxy.py]: typo: ot -> to

**Comment by jhawkesworth** [comment]
@jborean93 - a couple of tiny suggestions for the docs.  I can't comment usefully on the native code but the rest of it looks good to me. I haven't tested though.

**Comment by jborean93** [comment]
Thanks @jhawkesworth for the review, have made the changes.

**Comment by jborean93** [comment]
Thanks to everyone for the review.
